### ISSUE TITLE
Stacktrace contents getting illegal characters inserted with text function.

### ISSUE THREAD

**Issue by fj40bryan** [behave/behave#336]
Today we upgraded from 1.2.4 to 1.2.5 and we noticed that when a test fails our junit style reports contain invalid characters in the CDATA stacktraces.
We debugged it to this line:
model.py, line 1473 -> "error = _text(traceback.format_exc())"

The _text function seems to be new in this release and is mangling our stacktrace, inserting illegal characters.

If it is any extra value for reproduction, we are on Windows, with python 2.7.9.

## Here is a sample stack trace before it is run through the _text function:
Traceback (most recent call last):
  File "C:\\Users\\bXXXX\\Documents\\virtualenvs\\XXXX\\lib\\site-packages\\behave\\model.py", line 1456, in run
    match.run(runner.context)
  File "C:\\Users\\bXXXX\\Documents\\virtualenvs\\XXXX\\lib\\site-packages\\behave\\model.py", line 1903, in run
    self.func(context, *args, **kwargs)
  File "features\\steps\\installer_steps.py", line 210, in impl
    directories, task_names, reg_keys)
AssertionError

## Here is the resulting stacktrace, which when printed to the CDATA section of the report, produces an invalid report:

Traceback (most recent call last):
  File "C:�sersXXXX\Documentsirtualenvs\XXXX\lib\site-packagesehave\model.py", line 1456, in run
    match.run(runner.context)
  File "C:�sersXXXX\Documentsirtualenvs\XXXX\lib\site-packagesehave\model.py", line 1903, in run
    self.func(context, *args, **kwargs)
  File "features\steps\installer_steps.py", line 210, in impl
    directories, task_names, reg_keys)
AssertionError

We have worked-around the issue by patching calls to text and over-ridding the default encoding with "utf-8". So this is not a very important issue to see fixed, but we thought you would want to know. Hope this info helps! Bryan

**Comment by arsenik** [comment]
this issue is causing me a lot of trouble, any fix soon?

**Comment by mixxorz** [comment]
I'll try to work on this now.

**Comment by dpiche174** [comment]
I think the traceback should be decoded using the encoding returned by [sys.getfilesystemencoding()](https://docs.python.org/2/library/sys.html#sys.getfilesystemencoding) instead of utf-8: `_text(traceback.format_exc(), sys.getfilesystemencoding(), 'replace')` should be bulletproof.

Python tracebacks can contain mixed encoding (paths encoding depend of the OS and the error message encoding depends of the code that generates it).  By decoding the traceback with [sys.getfilesystemencoding()](https://docs.python.org/2/library/sys.html#sys.getfilesystemencoding), you are at least assured that the paths will be formatted correctly.

**Comment by lampmantech** [comment]
This problem is a major pain for us under Windows, as it provokes errors in our testrunner
in trying to give you the error traceback, and the reason for it is simple. It is NOT Windows specific, and has nothing to do with unicode or getfilesystemencoding().

If you look carefully at the traceback above after "an invalid report:" and you'll
notice that the non-encodable characters are generated after some (not all)
backslashes.  So \b in the string becomes a non-encodeable character when you
do something with that string in the _text function.

This is because you are taking a string with backslashes in it and then printing it to stdout/stderr
and it's parsing the string for \b as backspace and \f as formfeed etc - some are special but not all. This fault is generic to any Python program, and hypothetically would happen even under Unix if you had a filename in a backtrace that contained an embedded backslash, or in any other part of the traceback message. It's an outright error in the code and simple to fix.

The simplest fix is to simply double the backslashes, if there are any:
error = _text(traceback.format_exc().replace('\\','\\\\'))
All tracebacks on Windows can be expected to have backslashes followed by a letter,
and any of those sequences that are string escapes will always cause the problem.
Or you can use repr().

You are safer doing this on all platforms, because althought it's very rare, you can indeed
have  embedded backslashes in Unix too: 
cat > /tmp/'foo\\bar'
^D
ls /tmp/foo*
/tmp/foo\\bar

**Comment by lampmantech** [comment]
Whoops - github needs fixing too! It ate some, but not all, of my backslashes!!

All of the backslashes in the above post after "The simplest fix"
should be doubled. I assume github has this same bug too :-)

**Comment by dpiche174** [comment]
This issue is really a Unicode decoding issue.

In Python 2, `traceback.format_exc()` returns a bytestring. If you look at the function `behave.textutil.text()` which is called to convert the traceback into a unicode object, it uses the codec 'unicode-escape' by default. 'unicode-escape' causes \u and other backslash-escaped characters to be interpreted. 'unicode-escape' must only be used to decode bytestring containing Unicode escape codes, such as strings created by encoding a unicode object with parameter replace set to 'backslashreplace'.

IMO, I don't think that having a default encoding in `behave.textutil.text()` is a good idea. I would rather make it a required argument and update the code where it is used to pass the proper encoding depending on the context.

**Comment by jenisys** [comment]
This problem is currently on the top of my stack (especially for the Windows platform). A simple work-around is to replace the text() implementation: Replace "unicode-escape" with "UTF-8" or similar. This should help with most problems.

**Comment by lampmantech** [comment]
I think you're wrong and you misunderstand the problem. Study my comment again because
it is a exact solution, not something that "should help most". We've been running this fix for 2 months.

It has NOTHING to do with unicode escapes: the _text function is creating the problem
from an ANSI input that contains ANSI backslashes because of Python string interpolation: Python is creating things like \b a backspace, which is unprintable, out of the input
you are giving _text: Windows pathnames like \behave.

Either double all backslashes in the string to escape them (the most obvious solution in this use case), or wrap the argument in a repr().

**Comment by arsenik** [comment]
I am on windows and use repr() whenever I have an output containing a path.

**Comment by joshsleeper** [comment]
@jenisys, any news on this?

I could be mistaken, but #453, #449, and #396 all seem to be effectively this same problem, and it's giving my team a lot of grief right now.

**Comment by joshsleeper** [comment]
After reading this thread (thanks to @lampmantech and @dpiche174 for the legwork) and trying a few things myself, the solution that seems to work best for us is the following:

```diff
diff --git a/model.py b/traceback_friendly_model.py
index ca4d5ff..da0fd7d 100644
--- a/model.py
+++ b/traceback_friendly_model.py
@@ -1301,10 +1301,10 @@ class Step(BasicStatement, Replayable):
                 error = u"Assertion Failed: "+ message
             else:
                 # no assertion text; format the exception
-                error = _text(traceback.format_exc())
+                error = _text(traceback.format_exc(), sys.getfilesystemencoding())
         except Exception as e:      # pylint: disable=broad-except
             self.status = "failed"
-            error = _text(traceback.format_exc())
+            error = _text(traceback.format_exc(), sys.getfilesystemencoding())
             self.store_exception_context(e)

         self.duration = time.time() - start
```

@lampmantech's solution of doubling all backslashes does fix the errors, but also resulted in occasional cases of double backslashes being printed for me.

However, extending @dpiche174's solution and simply adding `sys.getfilesystemencoding()` as the second argument to the `_text()` calls for handling both `AssertionError`s **and** `Exception`s works like a charm for me though.

What do you think, @jenisys?
Does this seem like an acceptable solution?

**Comment by jenisys** [comment]
FIXED IN: behave-1.2.6.dev0
